### PR TITLE
Unpacking syntax closes #341

### LINKED ISSUE #341: Syntax for unpacking arrays
Currently it is already possible to enumerate arrays using
```
for index, value in array {...}
```
Similarly it is possible to iterate over the key-value pairs of dictionaries and arguments.
```
for key, value in dict {...}
```
To me this behaviour is not obvious. I think a more general and more intuitive way to do this is using an unpacking syntax, which is used in many modern programming languages. Some examples of how the new syntax might look:
```
let a, b = (1, 2) // a = 1, b = 2
let a, _, c = (1, 2, 3) // a = 1, c = 3

let a, b = (1, 2, 3) // error (too many values to unpack)
let a, b, c = (1, 2) // error (not enough values to unpack)

let a, ..b = (1, 2, 3) // a = 1, b = (2, 3)
let ..a, b = (1, 2, 3) // a = (1, 2), b = 3, not sure if this should be allowed
```
This would work the same with for-loops.

The examples at the beginning would then become
```
for index, value in array.enumerate() {...}
for key, value in dict.pairs() {...}
```

### PR DESCRIPTION
**Pull request by Marmare314**
This PR should work as intended, but the implementation might still be a bit unclean in some places. But I would be happy for any feedback on it.

- [x] Write documentation
- [x] Write tests
- [x] Test completion
- [x] Adopt new syntax in existing tests

### CODE REVIEW

File: docs/src/reference/scripting.md [hunk 1]
```diff
@@ -81,6 +81,19 @@ It explains #name.
...
 
+Let bindings can also be used to unpack arrays. Note that the underscore `_`
+is special in these assignments, as it can be used multiple times in the
+same assignment.
+
+```example
+#let (x, y) = (1, 2)
+The coordinates are #x, #y.
+
+#let (_, c1, .._, c2) = "Hello World!".clusters()
```
Comment (laurmaedje) [inline docs/src/reference/scripting.md:92]: would be nicer if `.._` could be just `..` I think. the same also applies to argument lists, where it currently doesn't work to but I guess that's my fault.

File: docs/src/reference/scripting.md [hunk 2]
```diff
@@ -81,6 +81,19 @@ It explains #name.
 Sum is #add(2, 3).
 ```
 
+Let bindings can also be used to unpack arrays. Note that the underscore `_`
```
Comment (laurmaedje) [inline docs/src/reference/scripting.md:84]: Would it be feasible to extend this to dictionaries and the binding names would have to exactly match the keys? Also: Is the underscore really special? Does `#let _ = 10` not do the same thing?

File: docs/src/reference/scripting.md [hunk 3]
```diff
@@ -136,20 +149,12 @@ For loops can iterate over a variety of collections:
...
-  `{for name, value in args {..}}` \
-  Iterates over the values or names and values of the
-  [arguments]($type/arguments). For positional arguments, the `name` is
-  `{none}`.
+  Iterates over the items in the [array]($type/array). The unpacking syntax
+  described in [Let binding]($scripting/bindings) can also be used in place here.
+
+- `{for pair in dict {..}}` \
+  Iterates over the key-value pairs of the [dictionary]($type/dictionary).
+  The pairs can also be unpacked by using `{for (key, value) in dict {..}}`
```
Comment (laurmaedje) [inline docs/src/reference/scripting.md:157]:
```suggestion
  The pairs can be unpacked by using `{for (key, value) in dict {..}}`
```

File: docs/src/reference/scripting.md [hunk 4]
```diff
@@ -136,20 +149,12 @@ For loops can iterate over a variety of collections:
...
-  Iterates over the values or keys and values of the
-  [dictionary]($type/dictionary).
-
-- `{for value in args {..}}` \
-  `{for name, value in args {..}}` \
-  Iterates over the values or names and values of the
-  [arguments]($type/arguments). For positional arguments, the `name` is
-  `{none}`.
+  Iterates over the items in the [array]($type/array). The unpacking syntax
+  described in [Let binding]($scripting/bindings) can also be used in place here.
```
Comment (laurmaedje) [inline docs/src/reference/scripting.md:153]:
```suggestion
  described in [Let binding]($scripting/bindings) can also be used here.
```

File: src/eval/array.rs [hunk 5]
```diff
@@ -322,6 +322,19 @@ impl Array {
...
+    pub fn enumerate(&self) -> Self {
+        let v = self
+            .0
+            .iter()
+            .enumerate()
+            .map(|(i, value)| eco_vec![Value::Int(i as i64), value.clone()])
+            .map(Self::from_vec)
+            .map(Value::Array)
+            .collect();
+        Self::from_vec(v)
```
Comment (laurmaedje) [inline src/eval/array.rs:336]: This should work (more or less, I can't really check from the web interface):
```suggestion
        self
            .iter()
            .enumerate()
            .map(|(i, value)| array![i, value.clone()])
            .collect()
```

File: src/eval/mod.rs [hunk 6]
```diff
@@ -1184,6 +1184,51 @@ impl Eval for ast::Closure {
...
+impl ast::Pattern {
+    // Destruct the given value into the pattern.
+    pub fn define(&self, vm: &mut Vm, value: Value) -> SourceResult<Value> {
+        match self.bindings() {
+            ast::PatternKind::Ident(ident) => {
+                vm.define(ident, value);
+                Ok(Value::None)
+            }
+            ast::PatternKind::Destructure(pattern) => {
+                let Ok(value) = value.clone().cast::<Array>() else {
```
Comment (laurmaedje) [inline src/eval/mod.rs:1196]: is this clone necessary?

File: docs/src/reference/scripting.md (hunk 1 above)
Comment (laurmaedje) [reply docs/src/reference/scripting.md:92]: by the way, this is really cool. I think argument sinks in the middle like this destructuring sink are actually broken because I couldn't be bothered to implement them properly...

File: src/eval/mod.rs [hunk 7]
```diff
@@ -1184,6 +1184,51 @@ impl Eval for ast::Closure {
...
+                        ast::DestructuringKind::Ident(ident) => {
+                            let Ok(v) = value.at(i) else {
+                                bail!(ident.span(), "not enough elements to destructure");
+                            };
+                            vm.define(ident.clone(), v.clone());
+                            i += 1;
+                        }
+                        ast::DestructuringKind::Sink(ident) => {
+                            let sink_size = i64::try_from(value.len()).unwrap()
+                                - i64::try_from(pattern.len()).unwrap()
```
Comment (laurmaedje) [inline src/eval/mod.rs:1212]: if you're unwrapping anyway might as well use `as` in my opinion. but I know that opinions on `as` differ.

File: src/eval/mod.rs [hunk 8]
```diff
@@ -1184,6 +1184,51 @@ impl Eval for ast::Closure {
...
+                                + 1;
+                            let Ok(v) = value.slice(i, Some(i + sink_size)) else {
```
Comment (laurmaedje) [inline src/eval/mod.rs:1214]: note that `slice` allows negative indices to index from the back. does this interfere with negative sink size?

File: src/eval/str.rs [hunk 9]
```diff
@@ -297,6 +297,11 @@ impl Str {
 
         Ok(resolved)
     }
+
+    /// Enumerate the grapheme clusters of this string.
+    pub fn enumerate(&self) -> Array {
+        self.clusters().enumerate()
+    }
```
Comment (laurmaedje) [inline src/eval/str.rs:304]: not sure about this one. I think an explicit `.clusters().enumerate()` in Typst code would be better since there are other things you might want to enumerate (e.g. codepoints)

File: src/syntax/ast.rs [hunk 10]
```diff
@@ -1533,7 +1533,13 @@ impl Closure {
     ///
     /// This only exists if you use the function syntax sugar: `let f(x) = y`.
     pub fn name(&self) -> Option<Ident> {
-        self.0.children().next()?.cast()
+        match self.0.cast_first_match::<Pattern>() {
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1536]: could this use the ? operator?

File: src/syntax/ast.rs [hunk 11]
```diff
@@ -1589,29 +1595,110 @@ impl AstNode for Param {
...
+    Sink(Ident),
+}
+
+#[derive(Debug, Clone, Hash)]
+pub enum PatternKind {
+    /// A single identifier: `x`.
+    Ident(Ident),
+    /// A destructuring pattern: `(x, _, ..y)`.
+    Destructure(Vec<DestructuringKind>),
+}
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1617]: could you flip the order and add documentation (with an example similar to the others)?

File: src/syntax/ast.rs [hunk 12]
```diff
@@ -1589,29 +1595,110 @@ impl AstNode for Param {
...
+    /// A destructuring pattern: `x` or `(x, _, ..y)`.
+    Pattern
+}
+
+#[derive(Debug, Clone, Hash)]
+pub enum DestructuringKind {
+    /// An identifier: `x`.
+    Ident(Ident),
+    /// An argument sink: `..y`.
+    Sink(Ident),
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1608]: this would then be `Sink(Option<Ident>)` if we allow just `(.., last)`

File: src/syntax/ast.rs [hunk 13]
```diff
@@ -1589,29 +1595,110 @@ impl AstNode for Param {
...
+                        .push(DestructuringKind::Sink(child.cast_first_match().unwrap())),
+                    _ => {}
+                }
+            }
+            PatternKind::Destructure(bindings)
+        } else if let Some(ident) = self.0.cast_first_match::<Ident>() {
+            PatternKind::Ident(ident)
+        } else {
+            PatternKind::Ident(Ident::default())
+        }
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1639]:
```suggestion
        } else {
            PatternKind::Ident(self.0.cast_first_match().unwrap_or_default())
        }
```

File: src/syntax/ast.rs [hunk 14]
```diff
@@ -1589,29 +1595,110 @@ impl AstNode for Param {
...
+            LetBindingKind::Normal(pattern)
+        } else if let Some(closure) = self.0.cast_first_match::<Closure>() {
+            if let Some(ident) = closure.name() {
+                LetBindingKind::Closure(ident)
+            } else {
+                LetBindingKind::Closure(Ident::default())
+            }
+        } else {
+            LetBindingKind::Normal(Pattern::default())
         }
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1694]: can this be written simpler like above? if not, oh well, but it feels like it can.

File: src/syntax/parser.rs [hunk 15]
```diff
@@ -833,22 +833,59 @@ fn args(p: &mut Parser) {
...
+// TODO (Marmare): remove success return value
+enum PatternKind {
+    Normal,
+    Destructuring,
+}
+
+fn pattern(p: &mut Parser) -> PatternKind {
+    let m = p.marker();
+    let unpacking = p.at(SyntaxKind::LeftParen);
+    if unpacking {
```
Comment (laurmaedje) [inline src/syntax/parser.rs:845]: I think its clear from the code that follows. Also, it would be nice if the terminology was clear: Either "unpacking" or "destructuring" (across all code, errors, and docs). I think I'm in favor of "destructuring".
```suggestion
    if p.at(SyntaxKind::LeftParen) {
```

File: src/syntax/parser.rs [hunk 16]
```diff
@@ -833,22 +833,59 @@ fn args(p: &mut Parser) {
...
+        collection(p, false);
+        validate_unpacking(p, m);
+        p.wrap(m, SyntaxKind::Pattern);
+
+        PatternKind::Destructuring
+    } else {
+        let success = p.expect(SyntaxKind::Ident);
+        if p.at(SyntaxKind::Comma) {
+            // TODO: this should only be a warning instead
+            p.expected("keyword `in`. Did you mean to use an unpacking pattern?");
```
Comment (laurmaedje) [inline src/syntax/parser.rs:855]: error messages should be all lowercase

File: src/syntax/parser.rs [hunk 17]
```diff
@@ -833,22 +833,59 @@ fn args(p: &mut Parser) {
...
+            if closure {
+                let m3 = p.marker();
+                collection(p, false);
+                validate_params(p, m3);
+                p.wrap(m3, SyntaxKind::Params);
+            }
+        }
+        PatternKind::Destructuring => {
+            destructuring = true;
+        }
```
Comment (laurmaedje) [inline src/syntax/parser.rs:885]:
```suggestion
        PatternKind::Destructuring => destructuring = true,
```

File: src/syntax/parser.rs [hunk 18]
```diff
@@ -1086,6 +1113,41 @@ fn validate_args(p: &mut Parser, m: Marker) {
...
+                let Some(within) = child.children_mut().last_mut() else { continue };
+                if within.kind() != SyntaxKind::Ident {
+                    within.convert_to_error(eco_format!(
+                        "expected identifier, found {}",
+                        within.kind().name(),
+                    ));
+                    child.make_erroneous();
+                }
+                if used_spread {
+                    child.convert_to_error("multiple argument sinks");
```
Comment (laurmaedje) [inline src/syntax/parser.rs:1136]:
```suggestion
                    child.convert_to_error("multiple destructuring sinks");
```
maybe reword to e.g. "at most one destructuring sink is allowed". if args has the same message, reword that, too.

**Review by laurmaedje — COMMENTED** [review]
overall, I'm very happy with how this turned out! I wanted to have this too and now I don't have to implement it. :)

File: src/eval/mod.rs (hunk 6 above)
Comment (Marmare314) [reply src/eval/mod.rs:1196]: It might not be, but this is the only way I got it to compile. In the else branch `value.type_name()` is called and otherwise it complains that a moved value is borrowed.

File: src/eval/mod.rs (hunk 7 above)
Comment (Marmare314) [reply src/eval/mod.rs:1212]: Thanks, I only learned about `as` after writing this and forgot about it.

File: docs/src/reference/scripting.md (hunk 2 above)
Comment (Marmare314) [reply docs/src/reference/scripting.md:84]: That seems like a useful addition. You're suggesting something like this right?
```
#let (a: x, c: y, ..rest) = ("a": 1, "b": 2, "c": 3, "d": 4)
#x // 1
#y // 2
#rest // ("c": 3, "d": 4)

#let (a: x) = ("z": 1) // error
```
I haven't looked into it, but I think that shouldn't be too hard to implement.

Alright, so the wording there definitely needs improvement. What I meant is that you can do
```
#let (_, _, _) = (1, 2, 3)
#_ // 3
```
But not
```
#let (a, a, a) = (1, 2, 3) // error
```

File: src/syntax/ast.rs (hunk 10 above)
Comment (Marmare314) [reply src/syntax/ast.rs:1536]: I updated this, but it feels like there is still some room for improvement, I'm just not sure how.

File: docs/src/reference/scripting.md (hunk 2 above)
Comment (laurmaedje) [reply docs/src/reference/scripting.md]: Yep, that's what I'm suggesting. And `a: a` would be the same as `a`.

File: src/syntax/ast.rs (hunk 14 above)
Comment (Marmare314) [reply src/syntax/ast.rs:1693]: Should be a lot cleaner now.

File: src/syntax/ast.rs (hunk 11 above)
Comment (Marmare314) [reply src/syntax/ast.rs]: I'm not quite sure what examples you mean, I have added some basic documentation for now. Also are the derives here necessary? The code compiles just fine without it.

File: src/syntax/ast.rs [hunk 19]
```diff
@@ -1589,29 +1593,121 @@ impl AstNode for Param {
...
+
+/// The kind of an element in a destructuring pattern.
+#[derive(Debug, Clone, Hash)]
+pub enum DestructuringKind {
+    /// An identifier: `x`.
+    Ident(Ident),
+    /// An argument sink: `..y`.
+    Sink(Option<Ident>),
+    /// Named arguments: `x: 1`.
+    Named((Ident, Ident)),
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1618]:
```suggestion
    Named(Ident, Ident),
```

File: src/syntax/ast.rs [hunk 20]
```diff
@@ -1589,29 +1593,121 @@ impl AstNode for Param {
...
+                        let ident = filtered_children.next().unwrap_or_default();
+                        bindings.push(DestructuringKind::Named((key, ident)))
+                    }
+                    _ => (),
+                }
+            }
+            PatternKind::Destructure(bindings)
+        } else {
+            PatternKind::Ident(self.0.cast_first_match().unwrap())
+        }
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1647]:
```suggestion
        if self.0.children().len() <= 1 {
            return PatternKind::Ident(self.0.cast_first_match().unwrap_or_default());
        }
        
        let mut bindings = Vec::new();
        for child in self.0.children() {
            match child.kind() {
                SyntaxKind::Ident => {
                    bindings.push(DestructuringKind::Ident(child.cast().unwrap_or_default()));
                }
                SyntaxKind::Spread => {
                    bindings.push(DestructuringKind::Sink(child.cast_first_match()));
                }
                SyntaxKind::Named => {
                    let mut filtered = child.children().filter_map(SyntaxNode::cast);
                    let key = filtered.next().unwrap_or_default();
                    let ident = filtered.next().unwrap_or_default();
                    bindings.push(DestructuringKind::Named(key, ident));
                }
                _ => {}
            }
        }
        
        PatternKind::Destructure(bindings)
```

File: src/syntax/ast.rs [hunk 21]
```diff
@@ -1533,7 +1533,11 @@ impl Closure {
     ///
     /// This only exists if you use the function syntax sugar: `let f(x) = y`.
     pub fn name(&self) -> Option<Ident> {
-        self.0.children().next()?.cast()
+        if let PatternKind::Ident(ident) = self.0.cast_first_match::<Pattern>()?.kind() {
+            Some(ident)
+        } else {
+            Option::None
+        }
```
Comment (laurmaedje) [inline src/syntax/ast.rs:1540]: `match` would be simpler here I think

File: src/eval/mod.rs [hunk 22]
```diff
@@ -1184,6 +1185,102 @@ impl Eval for ast::Closure {
...
+                                }
+                                ast::DestructuringKind::Named((key, _)) => {
+                                    bail!(
+                                        key.span(),
+                                        "cannot destructure named elements from an array"
+                                    );
+                                }
+                            }
+                        }
+                        if i < value.len().try_into().unwrap() {
```
Comment (laurmaedje) [inline src/eval/mod.rs:1234]: can also use `as`

File: src/eval/mod.rs [hunk 23]
```diff
@@ -1184,6 +1185,102 @@ impl Eval for ast::Closure {
...
+                                            "not enough elements to destructure"
+                                        );
+                                    }
+                                    let Ok(v) = value.slice(i, Some(i + sink_size)) else {
+                                        bail!(self.span(), "not enough elements to destructure");
+                                    };
+                                    if let Some(ident) = ident {
+                                        vm.define(ident.clone(), v.clone());
+                                    }
+                                    i += sink_size;
```
Comment (laurmaedje) [inline src/eval/mod.rs:1224]: by doing something like `value.len().checked_sub(pattern.len()).and_then(|c| value.slice(i, Some(i + c as i64)))` you can get rid of the duplicated error message

File: src/eval/mod.rs [hunk 24]
```diff
@@ -38,6 +38,7 @@ pub use self::value::*;
 pub(crate) use self::methods::methods_on;
 
 use std::collections::BTreeMap;
+use std::collections::HashSet;
```
Comment (laurmaedje) [inline src/eval/mod.rs:41]: these two imports can be merged

File: docs/src/reference/scripting.md [hunk 25]
```diff
@@ -81,6 +81,36 @@ It explains #name.
...
+  "Shakespeare": "Hamlet",
+  "Homer": "The Odyssey",
+  "Dicken": "Oliver Twist",
+  "Austen": "Persuasion",
+)
+#for (author, title) in books [
+  #author wrote #title, 
+]
+Homer wrote #Homer and Austen wrote #jane-austen.
+```
```
Comment (laurmaedje) [inline docs/src/reference/scripting.md:104]: maybe split this up into two smaller exmaples, otherwise it's gonna be a bit overwhelming. also ideally, the lines would be a bit shorter so that the docs can be side-by-side, but that's hard to tell without the ability to build the docs. I can also do it myself at the end.

**Review by laurmaedje — COMMENTED** [review]
A few more suggestions. By the way, it's super refreshing to get a PR with a comprehensive test suite. :)

File: docs/src/reference/scripting.md (hunk 25 above)
Comment (Marmare314) [reply docs/src/reference/scripting.md]: I updated the docs, but the examples are still a bit clunky. I'll try to think of something better.

**Comment by laurmaedje** [comment]
Thank you!
